File: npm/src/app/notifications.ts
import { on } from "./events.ts";
import { CefariError } from "./errors.ts";
import { invokeNotification } from "./results.ts";
import type {
  ActiveNotification,
  NotificationCapabilities,
  NotificationCategory,
  NotificationMediaReference,
  NotificationResponseEvent,
  NotificationResult,
  NotificationSendRequest,
  NotificationXdgCategory,
} from "./ipc.ts";
import type { Unsubscribe } from "./transport.ts";

export type NotificationPermission = { allowed: boolean };
export type NotificationSent = { id: string };
export type NotificationCategoriesRegistered = { count: number };
export type NotificationRemoved = { count: number };

export type SendNotificationInput = {
  title: string;
  body?: string | null;
  subtitle?: string | null;
  image?: NotificationMediaReference | null;
  icon?: NotificationMediaReference | null;
  iconRoundCrop?: boolean;
  threadId?: string | null;
  categoryId?: string | null;
  userInfo?: Record<string, string>;
  xdgCategory?: NotificationXdgCategory | null;
};

export type NotificationsApi = {
  permissionState(): Promise<NotificationPermission>;
  requestPermission(): Promise<NotificationPermission>;
  capabilities(): Promise<NotificationCapabilities>;
  registerCategories(
    categories: NotificationCategory[],
  ): Promise<NotificationCategoriesRegistered>;
  send(input: SendNotificationInput): Promise<NotificationSent>;
  active(): Promise<ActiveNotification[]>;
  removeDelivered(ids: string[]): Promise<NotificationRemoved>;
  removeAllDelivered(): Promise<NotificationRemoved>;
  onResponse(handler: (event: NotificationResponseEvent) => void): Unsubscribe;
};

export const notifications: NotificationsApi = {
  permissionState: async (): Promise<NotificationPermission> => {
    const result = await invokeNotification({
      command: "notification",
      payload: { notification: "permissionState" },
    });
    return permissionFromResult(result);
  },
  requestPermission: async (): Promise<NotificationPermission> => {
    const result = await invokeNotification({
      command: "notification",
      payload: { notification: "requestPermission" },
    });
    return permissionFromResult(result);
  },
  capabilities: async (): Promise<NotificationCapabilities> => {
    const result = await invokeNotification({
      command: "notification",
      payload: { notification: "capabilities" },
    });
    return capabilitiesFromResult(result);
  },
  registerCategories: async (
    categories: NotificationCategory[],
  ): Promise<NotificationCategoriesRegistered> => {
    const result = await invokeNotification({
      command: "notification",
      payload: {
        notification: "registerCategories",
        payload: { categories },
      },
    });
    return categoriesRegisteredFromResult(result);
  },
  send: async (input: SendNotificationInput): Promise<NotificationSent> => {
    const payload: NotificationSendRequest = {
      title: input.title,
      body: input.body ?? null,
      subtitle: input.subtitle ?? null,
      image: input.image ?? null,
      icon: input.icon ?? null,
      iconRoundCrop: input.iconRoundCrop ?? false,
      threadId: input.threadId ?? null,
      categoryId: input.categoryId ?? null,
      userInfo: input.userInfo ?? {},
      xdgCategory: input.xdgCategory ?? null,
    };
    const result = await invokeNotification({
      command: "notification",
      payload: { notification: "send", payload },
    });
    if (result.result === "sent") return result.payload;
    return unsupportedNotificationResult(result);
  },
  active: async (): Promise<ActiveNotification[]> => {
    const result = await invokeNotification({
      command: "notification",
      payload: { notification: "active" },
    });
    if (result.result === "active") return result.payload.notifications;
    return unexpectedNotificationResult(result, "active notifications");
  },
  removeDelivered: async (ids: string[]): Promise<NotificationRemoved> => {
    const result = await invokeNotification({
      command: "notification",
      payload: { notification: "removeDelivered", payload: { ids } },
    });
    return removedFromResult(result);
  },
  removeAllDelivered: async (): Promise<NotificationRemoved> => {
    const result = await invokeNotification({
      command: "notification",
      payload: { notification: "removeAllDelivered" },
    });
    return removedFromResult(result);
  },
  onResponse: (
    handler: (event: NotificationResponseEvent) => void,
  ): Unsubscribe => on("notification.response", handler),
};

function permissionFromResult(
  result: NotificationResult,
): NotificationPermission {
  switch (result.result) {
    case "permissionState":
    case "permissionRequested":
      return result.payload;
    case "permissionDenied":
      return { allowed: false };
    case "unsupported":
      throw new CefariError({
        code: "unsupported",
        details: {
          command: "notification",
          reason: result.payload.reason,
        },
      });
    case "sent":
    case "capabilities":
    case "categoriesRegistered":
    case "active":
    case "removed":
      throw new CefariError({
        code: "invalidCommand",
        details: {
          message: `unexpected notification result: ${result.result}`,
        },
      });
  }
}

function capabilitiesFromResult(
  result: NotificationResult,
): NotificationCapabilities {
  if (result.result === "capabilities") return result.payload;
  return unexpectedNotificationResult(result, "notification capabilities");
}

function categoriesRegisteredFromResult(
  result: NotificationResult,
): NotificationCategoriesRegistered {
  if (result.result === "categoriesRegistered") return result.payload;
  return unexpectedNotificationResult(result, "registered notification categories");
}

function removedFromResult(result: NotificationResult): NotificationRemoved {
  if (result.result === "removed") return result.payload;
  return unexpectedNotificationResult(result, "removed notifications");
}

function unsupportedNotificationResult(
  result: NotificationResult,
): never {
  if (result.result === "unsupported") {
    throw new CefariError({
      code: "unsupported",
      details: {
        command: "notification.send",
        reason: result.payload.reason,
      },
    });
  }
  if (result.result === "permissionDenied") {
    throw new CefariError({
      code: "denied",
      details: {
        message: "notification permission denied",
      },
    });
  }
  return unexpectedNotificationResult(result, "sent notification");
}

function unexpectedNotificationResult(
  result: NotificationResult,
  expected: string,
): never {
  throw new CefariError({
    code: "invalidCommand",
    details: {
      message: `expected ${expected}; received notification result ${result.result}`,
    },
  });
}

File: crates/cefari-core/src/ipc/notifications.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use specta::Type;

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(tag = "notification", content = "payload", rename_all = "camelCase")]
pub enum NotificationCommand {
    PermissionState,
    RequestPermission,
    Capabilities,
    RegisterCategories(NotificationRegisterCategoriesRequest),
    Send(NotificationSendRequest),
    Active,
    RemoveDelivered(NotificationRemoveDeliveredRequest),
    RemoveAllDelivered,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct NotificationSendRequest {
    pub title: String,
    pub body: Option<String>,
    pub subtitle: Option<String>,
    pub image: Option<NotificationMediaReference>,
    pub icon: Option<NotificationMediaReference>,
    pub icon_round_crop: bool,
    pub thread_id: Option<String>,
    pub category_id: Option<String>,
    pub user_info: BTreeMap<String, String>,
    pub xdg_category: Option<NotificationXdgCategory>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(tag = "source", content = "path", rename_all = "camelCase")]
pub enum NotificationMediaReference {
    AppResource(String),
    AppData(String),
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub enum NotificationXdgCategory {
    Call,
    CallEnded,
    CallIncoming,
    CallUnanswered,
    Device,
    DeviceAdded,
    DeviceError,
    DeviceRemoved,
    Email,
    EmailArrived,
    EmailBounced,
    Im,
    ImError,
    ImReceived,
    Network,
    NetworkConnected,
    NetworkDisconnected,
    NetworkError,
    Presence,
    PresenceOffline,
    PresenceOnline,
    Transfer,
    TransferComplete,
    TransferError,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct NotificationRegisterCategoriesRequest {
    pub categories: Vec<NotificationCategory>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct NotificationCategory {
    pub id: String,
    pub actions: Vec<NotificationCategoryAction>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(
    tag = "type",
    rename_all = "camelCase",
    rename_all_fields = "camelCase"
)]
pub enum NotificationCategoryAction {
    Action {
        id: String,
        title: String,
    },
    TextInput {
        id: String,
        title: String,
        input_button_title: String,
        input_placeholder: String,
    },
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct NotificationRemoveDeliveredRequest {
    pub ids: Vec<String>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct NotificationCapabilities {
    pub permission_state: bool,
    pub permission_prompt: bool,
    pub subtitle: bool,
    pub image: bool,
    pub icon: bool,
    pub icon_round_crop: bool,
    pub thread_id: bool,
    pub categories: bool,
    pub action_buttons: bool,
    pub text_input_actions: bool,
    pub user_info: bool,
    pub xdg_category: bool,
    pub active_notifications: bool,
    pub remove_delivered: bool,
    pub response_events: bool,
    pub cold_start_activation: bool,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct ActiveNotification {
    pub id: String,
    pub user_info: BTreeMap<String, String>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(tag = "result", content = "payload", rename_all = "camelCase")]
pub enum NotificationResult {
    PermissionState {
        allowed: bool,
    },
    PermissionRequested {
        allowed: bool,
    },
    Capabilities(NotificationCapabilities),
    CategoriesRegistered {
        count: u32,
    },
    Sent {
        id: String,
    },
    Active {
        notifications: Vec<ActiveNotification>,
    },
    Removed {
        count: u32,
    },
    PermissionDenied,
    Unsupported {
        reason: String,
    },
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(tag = "event", content = "payload", rename_all = "camelCase")]
pub enum NotificationEvent {
    Response(NotificationResponseEvent),
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct NotificationResponseEvent {
    pub id: String,
    pub action: NotificationAction,
    pub user_text: Option<String>,
    pub user_info: BTreeMap<String, String>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub enum NotificationAction {
    Default,
    Dismiss,
    Other(String),
}
